pkg/registries/google/transport.go

```diff
@@ -0,0 +1,57 @@
+package google
+
+import (
+	"net/http"
+
+	"github.com/heroku/docker-registry-client/registry"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/registries/docker"
+	"github.com/stackrox/rox/pkg/sync"
+	"golang.org/x/oauth2"
+)
+
+var log = logging.LoggerForModule()
+
+// googleTransport represents a transport that converts an oauth token source
+// into docker registry credentials.
+// This is required kind of trickery is required because the docker API does not
+// accept a standard oauth2 transport.
+type googleTransport struct {
+	registry.Transport
+	config      *docker.Config
+	token       *oauth2.Token
+	tokenSource oauth2.TokenSource
+	mutex       sync.Mutex
+}
+
+func newGoogleTransport(config *docker.Config, tokenSource oauth2.TokenSource) *googleTransport {
+	transport := &googleTransport{config: config, tokenSource: tokenSource}
+	if err := transport.refreshNoLock(); err != nil {
+		log.Error("Failed to refresh token: ", err)
+	}
+	return transport
+}
+
+func (t *googleTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	if !t.token.Valid() {
+		if err := t.refreshNoLock(); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
We can use `concurrency.WithLock` here _iff_ it's safe to do `t.Transport.RoundTrip()` outside of the lock - which looking at everything should be fine. This way we minimize the time we actually hold the lock.

Reply:
see follow up #9117